beacon-chain/blockchain/core.go

```diff
@@ -131,6 +146,43 @@ func (b *BeaconChain) CanProcessBlock(fetcher powchain.POWBlockFetcher, block *t
 	return validTime, nil
 }
 
+// CleanUpValidatorSet is called  every dynasty transition. It's primary function is
+// to go through queued validators and induct them to be active, and remove bad
+// active validator (balances below threshold) to the exit set. It also cross checks
+// every validator's switch dynasty before induct or remove.
+func (b *BeaconChain) CleanUpValidatorSet() ([]types.ValidatorRecord, []types.ValidatorRecord, []types.ValidatorRecord) {
+
+	var newExitedValidator []types.ValidatorRecord
+	newActiveValidators := b.state.CrystallizedState.ActiveValidators
+
+	// Loop through active validator set, remove validator whose balance is below 50% and switch dynasty > current dynasty.
+	for i, validator := range newActiveValidators {
+
+		if validator.Balance < params.DefaultBalance/2 {
+			newExitedValidator = append(newExitedValidator, validator)
+			newActiveValidators = append(newActiveValidators[:i], newActiveValidators[:i+1]...)
```

Comment:
don't think you can modify the slice while iterating over it: https://play.golang.org/p/wullBC6Q2CG

Reply:
i would suggest putting new validators into a new slice